performant_PoCs/focused-vision/focused_pipeline/utils/grid_layout_node.py

```diff
@@ -0,0 +1,146 @@
+import math
+from typing import List, Tuple
+import cv2
+import numpy as np
+import depthai as dai
+
+
+def _layout_rects(n: int) -> List[Tuple[float, float, float, float]]:
+    """
+    Returns a list of (x, y, w, h) tiles in normalized [0..1] coords.
+    Simple, deterministic layouts for 1..N items with up to 3 columns.
+    """
+    R: List[Tuple[float, float, float, float]] = []
+
+    def row(cols: int, y0: float, h: float, count: int | None = None):
+        if count is None:
+            count = cols
+        w = 1.0 / cols
+        for i in range(count):
+            R.append((i * w, y0, w, h))
+
+    if n <= 1:
+        return [(0.0, 0.0, 1.0, 1.0)]
+    if n == 2:
+        row(2, 0.0, 1.0, 2)
+    elif n == 3:
+        row(2, 0.0, 0.5, 2)
+        R.append((0.0, 0.5, 1.0, 0.5))
+    elif n == 4:
+        row(2, 0.0, 0.5, 2)
+        row(2, 0.5, 0.5, 2)
+    elif n == 5:
+        row(3, 0.0, 0.5, 3)
+        row(2, 0.5, 0.5, 2)
+    elif n == 6:
+        row(3, 0.0, 0.5, 3)
+        row(3, 0.5, 0.5, 3)
+    else:
+        rows = math.ceil(n / 3)
+        h = 1.0 / rows
+        left = n
+        y = 0.0
+        for _ in range(rows):
+            c = min(3, left)
+            row(c, y, h, c)
+            left -= c
+            y += h
+    return R
+
+
+def _fit_letterbox(img: np.ndarray, W: int, H: int) -> np.ndarray:
+    """Resize with preserved aspect and center into WxH canvas."""
+    ih, iw = img.shape[:2]
+    if iw <= 0 or ih <= 0:
+        return np.zeros((H, W, 3), dtype=np.uint8)
+
+    s = min(W / max(1, iw), H / max(1, ih))
+    nw, nh = max(1, int(iw * s)), max(1, int(ih * s))
+    resized = cv2.resize(img, (nw, nh), interpolation=cv2.INTER_AREA)
+
+    canvas = np.zeros((H, W, 3), dtype=np.uint8)
+    x0 = (W - nw) // 2
+    y0 = (H - nh) // 2
+    canvas[y0:y0 + nh, x0:x0 + nw] = resized
+    return canvas
+
+
+def _mosaic_from_frames(
+    frames: List[dai.ImgFrame],
+    out_w: int,
+    out_h: int,
+    frame_type: dai.ImgFrame.Type
+) -> dai.ImgFrame:
+    """
+    Compose a mosaic from a list of ImgFrames (assumed BGR/BGRA/GRAY).
+    Does not set timestamp/seq — caller should stamp from the GatherData reference.
```

Comment:
I don't see the caller would do that. Why do you say here the caller should do that but it doesn't?